### ISSUE TITLE
Better docs for navigation-2d user handlers ?

### ISSUE THREAD

**Issue by Vaguery** [quil/quil#328]
As I mentioned in https://github.com/quil/quil/issues/327, I'm trying to intercept `:mouse-dragged` and `:mouse-wheel` events in `:navigation-2d`, to filter out events that happen outside the sketch bounds.

I originally thought there was a bug in `:fun-mode`, simply because I didn't understand how it was working, given the docs.

But I confess that after several hours I still can't quite understand how event-handling is wrapped in the `:navigation-2d` hierarchy.

I've built this simple drag-handler:

``` cljs
(defn user-mouse-dragged
  [state event]
  (if (mouse-within? (:p-x event) (:p-y event))
    (do (prn "in") state)
    (do (prn "out") state)
    ))
```

And it does print "in" and "out" correctly to the fig wheel stdout stream.

What I'd _like_ to do though is block or intercept events that happen outside the draw bounds. But since the `mouse-dragged` event handler's signature is `[state event]->state`, I'm not sure how to do that, or even whether it's a good idea.

Any advice welcomed. I assume that I shouldn't be messing with the event object itself, even if I did know where it was located in the codebase.

Is there a way to filter events so that the default handlers in `navigation-2d` don't see them?

**Comment by heyarne** [comment]
You can `(assoc state :mouse-within? (mouse-within? (:p-x event) (:p-y event)))` to update your sketch accordingly in other functions. I'm not sure if I understand you correctly but the idea is that you keep everything that's relevant to how you draw in `state` and then read from it in `draw`. Hope this helps!

**Comment by nbeloglazov** [comment]
I believe Bill wants to "interact" with `:navigation-2d` middleware and disable it in some cases. I think currently it's not possible because in `navigation-2d` we first apply middleware's on-drag function only then user-provided function: https://github.com/quil/quil/blob/master/src/cljc/quil/middlewares/navigation_2d.cljc#L87

Though I don't think there was a good reason to use this specific order. Instead we could reverse it like: 
```clj
:mouse-dragged (fn [state event]
                 (mouse-dragged (user-mouse-dragged state event) event))
```

And add some check into `mouse-dragged` that disables if `:navigation-2d` contains field like `:skip-next-iteration` and that field can be set by the user-provided mouse-dragged handler.

**Comment by Vaguery** [comment]
Yes, sorry, didn't see the notifications. And yes: the problem I was having with `:navigation-2d` was that I was unable to interact with the surrounding web page: any scroll, drag or click needed to interact with a form on the page (for controlling the `quil` sketch args) would also scroll, leading to lost objects in an infinite field :)
